### ISSUE TITLE
Support for Robot Vacuum Cleaners in iOS 18

### ISSUE THREAD

**Issue by azmke** [homebridge/HAP-NodeJS#1073]
### Current Situation

Currently, robot vacuum cleaners are not natively supported by Homebridge or HAP-NodeJS. As a result, they are often represented as switches in existing plugins, which limits their functionality and integration within the Home app. According to Apple, robot vacuum cleaner support is expected to be natively available in the Home app with iOS 18, allowing them to participate in automations, scenes, and be controlled via Siri. However, I couldn't find any indication of support for vacuum robots in the code of either the [Homebridge](https://github.com/homebridge/homebridge) or the [HAP-NodeJS](https://github.com/homebridge/HAP-NodeJS) repositories.

### Proposed Change

Introduce native support for robot vacuum cleaners in HAP-NodeJS to leverage the new capabilities in iOS 18. Apple has announced native support for robot vacuum cleaners in iOS 18, which means Homebridge should enable vacuum cleaner plugins to utilize these new capabilities. This would involve adding a new accessory category and associated characteristics to properly represent robot vacuums. This change would then enable Homebridge to support plugins that can fully utilize the new robot vacuum cleaner controls, providing a more integrated and functional user experience.

### Additional Context

- Source: [Apple Newsroom - Updates to the Home Experience](https://www.apple.com/newsroom/2024/06/updates-to-the-home-experience-elevate-entertainment-and-bring-more-convenience/)
- Official Apple page for iOS 18 mentions that robot vacuum cleaner support is "coming later this year": [iOS 18](https://www.apple.com/ios/ios-18/)
- The implementation of this feature would allow users to integrate their robot vacuum cleaners more effectively into the Home app, enabling better automation and control through Siri.
- Example use case: A user could create a scene that includes starting the robot vacuum cleaner along with other actions, or use voice commands to control the vacuum cleaner directly.

**Comment by ebaauw** [comment]
We need to see whether these will actually be supported through HAP.  I suspect they might be Matter only.

**Comment by hjdhjd** [comment]
As @ebaauw indicated they're currently Matter-only, specifically Matter 1.2 which iOS 18 now supports. HAP-NodeJS is HomeKit-specific, at least at the moment. I wouldn't bet on this happening anytime soon I'm afraid. It remains to be seen whether or not Apple will ultimately deprecate HomeKit in favor of Matter or not over time.

TL;DR: unlikely to happen until **HomeKit** supports robot vacuums (or any other device class).

**Comment by adriancable** [comment]
@hjdhjd - iOS 18 only supports Matter 1.1 (with the sole, and odd, exception that it supports air purifiers from the Matter 1.2 spec) - no robot vacuums. iOS 18.1 as of the latest developer beta is the same. Maybe it will come in 18.2 later this year.

**Comment by azmke** [comment]
Hey together,

thanks to the three of you for the technical explanation. This makes me wonder about the strategic direction of Homebridge: As far as I understand, Homebridge only supports HomeKit at the moment. Given that HomeKit seems to be becoming less and less important and Matter is becoming more and more important, wouldn't it be wise to implement a Matter interface in the near future in order to benefit from new features (such as robot vacuum functions in this case)?

At the end of the day, the user doesn't care how the Homebridge communicates (HomeKit or Matter) with the home hub (HomePod etc.). They don't really notice anything about it. They just want their IOT device to be controllable via the Apple Home app and Siri. Which is kind of the core philosophy behind Homebridge, as far as I understand it.

What do you think?

**Comment by adriancable** [comment]
@azmke - HomeKit won't become less and less important until Matter supports most of the stuff that HomeKit does. Right now, Matter doesn't support some of the most common HomeKit device classes, like cameras. That will change eventually, but not soon.

There is matterbridge, which is a plug-ins wrapper around the official matter.js Matter implementation in JavaScript, which is essentially what you are asking for: https://github.com/Luligu/matterbridge

**Comment by hjdhjd** [comment]
> @hjdhjd - iOS 18 only supports Matter 1.1 (with the sole, and odd, exception that it supports air purifiers from the Matter 1.2 spec) - no robot vacuums. iOS 18.1 as of the latest developer beta is the same. Maybe it will come in 18.2 later this year.

Thanks for the clarification @adriancable - I must've missed it somewhere. 😄 

@azmke HomeKit's going to be supported for a very long time to come. I wouldn't count Apple out in one direction or the other...they like their own standards for things too. As to what Homebridge will support in the future, beyond HomeKit, we're always looking at the evolving smart home capability landscape and exploring accordingly. First, we'd like to get Homebridge 2.0 out the door. 😄

**Comment by afharo** [comment]
It looks like Apple is finally adding support for this: https://9to5mac.com/2024/12/16/ios-18-3-beta-adding-home-app-support-for-robot-vacuums/?extended-comments=1

Maybe it's worth reopening this issue?

**Comment by adriancable** [comment]
@afharo - what is the 'this' that Apple is finally adding support for? Support for Matter robot vacuums has no bearing on Homebridge or anything else HAP, unless Apple is also secretly adding HomeKit services and characteristics for robot vacuums which is possible, sure, but a bit of leap of hope without any supporting facts.

Also, the article is false/clickbait. What is true: someone on X reported finding strings referencing robot vacuum cleaners in iOS 18.3 beta 1. These strings are also there in iOS 18.2. Strings just mean it's being worked on, which we have already known since WWDC last June. That is all there is right now. The iOS 18.3 beta 1 Home app does not support Matter robot vacuums, despite what the article claims.

**Comment by afharo** [comment]
@adriancable, apologies for the (probably) false hopes. For a moment I was filled with hopium myself.

That being said, shouldn't we keep this issue open until we figure out if support is only coming for Matter or both (HomeKit and Matter), and, if the latter, then plan for action?

**Comment by hjdhjd** [comment]
@afharo It really doesn't. Issues should be open for things that we know exist...until it's released, it just adds to our maintenance workload. Every moment spent on rumors rather than reality is a moment that takes away from us writing code, improving the codebase, crushing bugs, and addressing our project priorities.

When something is actually released for **HomeKit** we will support it. Matter support is not currently part of our scope and there are other projects out there that implement Matter support.

Appreciate the enthusiasm - we all do. This is an actively developed project and will continue to be. We keep track of what Apple is doing and as new capabilities are released and make sense to implement, we will. 😀

**Comment by adriancable** [comment]
@hjdhjd - my personal argument for keeping this open is that it then remains something we can easily direct people to when they ask the question. A lot of people don't understand the difference between Matter and HomeKit, and I suspect if we close this, we'll just get other people opening 'issues' about the same thing which will suck up more of the community's time. But I also totally appreciate your argument to the contrary.

**Comment by hjdhjd** [comment]
@adriancable All for it...let's pin it and lock it perhaps?

**Comment by adriancable** [comment]
@hjdhjd - SGTM!

**Comment by adriancable** [comment]
@afharo, @hjdhjd et al. - iOS 18.4 beta 1 finally introduces support for Matter RVCs. But as expected, there are no parallel 'underlying' HomeKit services/characteristics. (for example, the RVC accessory appears under HMHome.accessories in an iOS app, but accessory.services is empty.)

Furthermore, iOS does include a 'map' between Matter clusters and HomeKit services/characteristics so that iOS apps using the HomeKit SDK can more or less 'transparently' work with Matter devices. This map does have some additions in iOS 18.4 beta 1 (for example, for smoke/CO detectors) but RVCs are not there.

So I think we can put to bed the 'hope' of RVC support in HAP-NodeJS.

**Comment by donavanbecker** [comment]
I would agree, I feel like we can close this and if anyone is wanting this support in Homebridge, we will need to add matter.js support to Homebridge.
